### PR TITLE
[UNR-1367] RPC tracking/metrics

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
RPC tracking is enabled with `SpatialStartRPCMetrics` and stopped with `SpatialStopRPCMetrics` console commands. It will record every sent RPC as well as the size of its payload, and then display tracked data upon stopping. Calling these console commands on the client will also start/stop RPC tracking on the server.

image

#### Documentation
Docs ticket created (UNR-1604)

#### Primary reviewers
@m-samiec @joshuahuburn

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 1]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+
+	UE_LOG(LogSpatialMetrics, Log, TEXT("Recording RPC metrics"));
+
+	bRPCTrackingEnabled = true;
+	RPCTrackingStartTime = FPlatformTime::Seconds();
+
+	// If RPC tracking is activated on a client, send a command to the server to start tracking.
+	if (!NetDriver->IsServer() && NetDriver->Sender != nullptr && NetDriver->GetWorld() != nullptr && NetDriver->GetWorld()->GetGameInstance() != nullptr)
+	{
+		FUnrealObjectRef PCObjectRef = NetDriver->PackageMap->GetUnrealObjectRefFromObject(NetDriver->GetWorld()->GetGameInstance()->GetPrimaryPlayerController());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: Any reasons not to use the NetConnection's owning actor here?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 2]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+{
+	if (!bRPCTrackingEnabled)
+	{
+		UE_LOG(LogSpatialMetrics, Log, TEXT("Haven't been recording RPC metrics"));
+		return;
+	}
+
+	// Display recorded sent RPCs.
+	const double TrackRPCInterval = FPlatformTime::Seconds() - RPCTrackingStartTime;
+	UE_LOG(LogSpatialMetrics, Log, TEXT("Recorded %d unique RPCs over the last %.3f seconds:"), RecentRPCs.Num(), TrackRPCInterval);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]:
```suggestion
	UE_LOG(LogSpatialMetrics, Log, TEXT("Recorded %d unique RPCs over the last %.3lf seconds:"), RecentRPCs.Num(), TrackRPCInterval);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 3]
```diff
@@ -590,6 +595,8 @@ void USpatialSender::SendRPC(TSharedRef<FPendingRPCParams> Params)
 					EntityId, CommandRequest.component_id, *Params->Function->GetName());
 			}
 		}
+		// TODO: Track queued RPCs.
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 4]
```diff
@@ -631,6 +640,13 @@ void USpatialSender::SendRPC(TSharedRef<FPendingRPCParams> Params)
 				Connection->SendComponentUpdate(EntityId, &ComponentUpdate);
 			}
 		}
+
+		if (!UnresolvedObject)
+		{
+			NetDriver->SpatialMetrics->TrackSentRPC(Params->Function, RPCInfo->Type, RPCPayloadSize);
+		}
+		// TODO: Track queued RPCs.
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -1029,6 +1032,24 @@ void USpatialReceiver::OnCommandRequest(Worker_CommandRequestOp& Op)
 		return;
 	}
 #endif // WITH_EDITOR
+	else if (Op.request.component_id == SpatialConstants::DEBUG_METRICS_COMPONENT_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Worth wrapping this and associated functionality in #if UE_BUILD_SHIPPING

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 6]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+		if (ControllerEntityId != SpatialConstants::INVALID_ENTITY_ID)
+		{
+			Worker_CommandRequest Request = {};
+			Request.component_id = SpatialConstants::DEBUG_METRICS_COMPONENT_ID;
+			Request.schema_type = Schema_CreateCommandRequest(SpatialConstants::DEBUG_METRICS_COMPONENT_ID, SpatialConstants::DEBUG_METRICS_START_RPC_METRICS_ID);
+			NetDriver->Connection->SendCommandRequest(ControllerEntityId, &Request, SpatialConstants::DEBUG_METRICS_START_RPC_METRICS_ID);
+		}
+		else
+		{
+			UE_LOG(LogSpatialMetrics, Warning, TEXT("SpatialStartRPCMetrics: Could not resolve local PlayerController entity!"));
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: If this happened should we reset bRPCTrackingEnabled and RPCTrackingStartTime?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 7]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+void USpatialMetrics::OnStartRPCMetricsCommand()
+{
+	SpatialStartRPCMetrics();
+}
+
+void USpatialMetrics::SpatialStopRPCMetrics()
+{
+	if (!bRPCTrackingEnabled)
+	{
+		UE_LOG(LogSpatialMetrics, Log, TEXT("Haven't been recording RPC metrics"));
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: "Could not stop recording RPC metrics. RPC Metrics not yet started."

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 8]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+		return;
+	}
+
+	// Display recorded sent RPCs.
+	const double TrackRPCInterval = FPlatformTime::Seconds() - RPCTrackingStartTime;
+	UE_LOG(LogSpatialMetrics, Log, TEXT("Recorded %d unique RPCs over the last %.3f seconds:"), RecentRPCs.Num(), TrackRPCInterval);
+
+	if (RecentRPCs.Num() > 0)
+	{
+		// NICELY log sent RPCs.
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: NICELY!

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 9]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+			TotalCalls += Stat.Calls;
+			TotalPayload += Stat.TotalPayload;
+		}
+		UE_LOG(LogSpatialMetrics, Log, TEXT("%s"), *SeparatorLine);
+		UE_LOG(LogSpatialMetrics, Log, TEXT("Total              | %s | %10d | %10.4f | %13d | %12.4f | %11.4f"), *FString::ChrN(MaxRPCNameLen, ' '), TotalCalls, TotalCalls / TrackRPCInterval, TotalPayload, (float)TotalPayload / TotalCalls, TotalPayload / TrackRPCInterval);
+
+		RecentRPCs.Empty();
+	}
+
+	bRPCTrackingEnabled = false;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: Reset the tracking time too?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 10]
```diff
@@ -60,3 +68,156 @@ double USpatialMetrics::CalculateLoad() const
...
+		if (ControllerEntityId != SpatialConstants::INVALID_ENTITY_ID)
+		{
+			Worker_CommandRequest Request = {};
+			Request.component_id = SpatialConstants::DEBUG_METRICS_COMPONENT_ID;
+			Request.schema_type = Schema_CreateCommandRequest(SpatialConstants::DEBUG_METRICS_COMPONENT_ID, SpatialConstants::DEBUG_METRICS_STOP_RPC_METRICS_ID);
+			NetDriver->Connection->SendCommandRequest(ControllerEntityId, &Request, SpatialConstants::DEBUG_METRICS_STOP_RPC_METRICS_ID);
+		}
+		else
+		{
+			UE_LOG(LogSpatialMetrics, Warning, TEXT("SpatialStopRPCMetrics: Could not resolve local PlayerController entity!"));
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: Should probably also mention this means we haven't been able to turn off tracking on the server.

**Comment by joshuahuburn** [comment]
Awesome stuff! Will there be some docs with this so users know it exists?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 6 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: I'm not sure if failing to send the request to the server should invalidate tracking this locally as well.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 9 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: We reset it when starting the tracking. It's not used in between.

**Pull request merged**